src/Microsoft.AspNetCore.Routing/DependencyInjection/DispatcherServiceCollectionExtensions.cs

```diff
@@ -43,11 +42,13 @@ public static IServiceCollection AddDispatcher(this IServiceCollection services)
             services.TryAddSingleton<IEndpointFinder<string>, NameBasedEndpointFinder>();
             services.TryAddSingleton<IEndpointFinder<RouteValuesBasedEndpointFinderContext>, RouteValuesBasedEndpointFinder>();
             services.TryAddSingleton<LinkGenerator, DefaultLinkGenerator>();
+
             //
             // Endpoint Selection
             //
-            services.TryAddSingleton<EndpointSelector>();
+            services.TryAddSingleton<EndpointSelector, EndpointConstraintEndpointSelector>();
```

Comment:
So **right now** we're using an implementation of `EndpointSelector` that supports `EndpointConstraint`. My hope is that we'd take this code to MVC as a back-compat feature, and delete `EndpointConstraint` from routing. 

As outlined in this PR we have options for policies that are an order of magnitude more performant than `EndpointConstraint`/`IActionConstraint`, and I'm aware of 4-5 actual implementations of `IActionConstraint` in the wild.

I had a hard think about what I want to optimize for scenario-wise. I want to make sure that our main path experiences like HTTP methods, consumes and versioning are really efficient. I'm willing to trade complexity for that. 

I think something that will help a lot would be a way to associate an `IRouteConstraint` with an endpoint/action in MVC. I think there's a lot of value in 'out of line' constraints that don't map to a route value for MVC, and we've never had that feature.